### PR TITLE
Use ZKSigma in stead of built in crypto.go

### PR DESCRIPTION
**Pull request by gertjaap**
This pull request removes crypto.go and replaces it with the use of [ZKSigma](https://github.com/mit-dci/zksigma)

In order for this PR to be merged, a couple updates to ZKSigma need to be merged in, that are in [ZKSigma PR #16](https://github.com/mit-dci/zksigma/pull/16)

### CODE REVIEW

**Review by Nabeelperson — COMMENTED** [review]
All good! Only a small error at line 126 in pki.go 
Print string needs a placeholder for `pk`.

File: audit.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package zkledger
+package apl
```
Comment (narula) [inline audit.go]: should be zkledger

File: bank_test.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package zkledger
+package apl
```
Comment (narula) [inline bank_test.go]: same

**Comment by gertjaap** [comment]
Okay, I fixed all the pending issues. First need to merge https://github.com/mit-dci/zksigma/pull/17 _and_ make it public, then we can merge this.

File: audit.go [hunk 3]
```diff
@@ -34,11 +36,11 @@ func MakeAuditor(num int, pki *PKI) *Auditor {
...
 		Setup:        make(chan struct{}),
 	}
-	a.CommsCache = make([]ECPoint, a.num)
-	a.RTokenCache = make([]ECPoint, a.num)
+	a.CommsCache = make([]zksigma.ECPoint, a.num)
+	a.RTokenCache = make([]zksigma.ECPoint, a.num)
 	for i := 0; i < a.num; i++ {
-		a.CommsCache[i] = EC.Zero()
-		a.RTokenCache[i] = EC.Zero()
+		a.CommsCache[i] = zksigma.Zero
```
Comment (narula) [inline audit.go:42]: how is this safe?  it means all the caches are using the same underlying big ints, right?

File: audit.go [hunk 4]
```diff
@@ -172,18 +174,18 @@ func (a *Auditor) computeSum(bank_i int) (*big.Int, bool) {
...
 			}
 		}
 	}
 	a.mu.Unlock()
-	gv := EC.G.Mult(rep.Sum).Neg() // 1 / g^\sum{v_i}
-	T := EC.Add(comms, gv)
-	verifies := VerifyEquivalence(T, rtokens, EC.H, a.pki.Get(bank_i), rep.Eproof)
+	gv := ZKLedgerCurve.Neg(ZKLedgerCurve.Mult(ZKLedgerCurve.G, rep.Sum)) // 1 / g^\sum{v_i}
+	T := ZKLedgerCurve.Add(comms, gv)
+	verifies, _ := rep.Eproof.Verify(ZKLedgerCurve, T, rtokens, ZKLedgerCurve.H, a.pki.Get(bank_i))
```
Comment (narula) [inline audit.go:189]: add a TODO for checking the error

File: transaction.go [hunk 5]
```diff
@@ -84,51 +86,64 @@ func (e *EncryptedTransaction) print_decrypted() {
...
-func (e *EncryptedTransaction) Verify(pks []ECPoint, CommCache []ECPoint, RTokenCache []ECPoint, debug string) bool {
+func (e *EncryptedTransaction) Verify(pks []zksigma.ECPoint, CommCache []zksigma.ECPoint, RTokenCache []zksigma.ECPoint, debug string) bool {
 	if e.skipVerify {
 		return true
 	}
-	g := GSPFS{curve: EC.C, ExponentPrime: EC.N, Generator: EC.H}
 	// Issuance
 	if e.Type == Issuance {
 		en := &e.Entries[e.Sender]
+		e.print_decrypted()
```
Comment (narula) [inline transaction.go]: is this supposed to be here?

File: transaction.go [hunk 6]
```diff
@@ -84,51 +86,64 @@ func (e *EncryptedTransaction) print_decrypted() {
...
 			Dprintf(" [%v] ETX %v Failed verify; issuance transaction values must be positive\n",
 				debug, e.Index)
 			return false
 		}
 		// Check proof of knowledge of sk_{asset issuer}
-		if !g.Verify(pks[len(pks)-1], en.SKProof) {
+		// TODO: Error handling
+		ok := false
+		if en.SKProof != nil {
+			ok, _ = en.SKProof.Verify(ZKLedgerCurve, pks[len(pks)-1])
```
Comment (narula) [inline transaction.go]: TODO to check for errors

File: audit.go (hunk 3 above)
Comment (gertjaap) [reply audit.go:42]: As far as I can tell, the only operation done on these cache points is Add. And an Add to Zero will return the point you're adding. But even an Add between two non-zero points returns new BigInts. So the Zero point will never be modified.

File: transaction.go (hunk 5 above)
Comment (Nabeelperson) [reply transaction.go]: I think this is an artifact of Willy's code. Probably just to see some terminal output of what is happening. Does not need to be passed around anywhere so I don't think it is a big issue.

**Pull request merged**